### ISSUE TITLE
[Bug]: Wrapping SnapbackZoom around Image doesn't work

### ISSUE THREAD

**Issue by Bauss9** [Glazzes/react-native-zoom-toolkit#57]
### Summary

I have a complex feed with carousel .etc and I implemented the SnapbackZoom around my image.

It did not work, the image disappeared as soon as I wrapped it with the SnapbackZoom, of course I changed a lot of stuff around, played with styles .etc, no results.

Then I did the most basic example in my test screen, with a simple static image of 100% width and 400px height, and wrapped it in SnapbackZoom, same thing again, the image disappears and is not showing. Causes?

### Environment
```json
{
  "name": "boooooom",
  "main": "expo-router/entry",
  "version": "1.0.0",
  "scripts": {
    "start": "expo start --dev-client",
    "reset-project": "node ./scripts/reset-project.js",
    "android": "expo run:android",
    "ios": "expo run:ios",
    "web": "expo start --web",
    "test": "jest --watchAll",
    "lint": "expo lint"
  },
  "dependencies": {
    "@10play/tentap-editor": "^0.5.21",
    "@expo/config": "^8.1.1",
    "@expo/metro-config": "^0.10.0",
    "@expo/vector-icons": "^14.0.2",
    "@gorhom/bottom-sheet": "^4.6.4",
    "@react-native-async-storage/async-storage": "^1.24.0",
    "@react-native-community/cli-platform-ios": "^14.0.0",
    "@react-native-community/datetimepicker": "^8.2.0",
    "@react-native-community/slider": "^4.5.2",
    "@react-native-picker/picker": "^2.8.0",
    "@react-native/metro-config": "^0.74.87",
    "@react-navigation/native": "^6.0.2",
    "@react-navigation/stack": "^6.4.1",
    "@shopify/react-native-skia": "^1.3.10",
    "axios": "^1.7.3",
    "date-fns": "^3.6.0",
    "expo": "~51.0.26",
    "expo-av": "^14.0.6",
    "expo-blur": "~13.0.2",
    "expo-camera": "~15.0.14",
    "expo-clipboard": "^6.0.3",
    "expo-constants": "~16.0.2",
    "expo-crypto": "~13.0.2",
    "expo-dev-client": "~4.0.23",
    "expo-font": "~12.0.9",
    "expo-image": "~1.12.14",
    "expo-image-manipulator": "^12.0.5",
    "expo-image-picker": "^15.0.7",
    "expo-linear-gradient": "~13.0.2",
    "expo-linking": "~6.3.1",
    "expo-location": "^17.0.1",
    "expo-random": "~14.0.1",
    "expo-router": "~3.5.21",
    "expo-splash-screen": "~0.27.5",
    "expo-status-bar": "~1.12.1",
    "expo-system-ui": "~3.0.7",
    "expo-updates": "~0.25.22",
    "expo-web-browser": "~13.0.3",
    "jwt-decode": "^4.0.0",
    "lottie-react-native": "6.7.0",
    "lucide-react-native": "^0.427.0",
    "react": "18.2.0",
    "react-dom": "18.2.0",
    "react-native": "0.74.5",
    "react-native-animated-dots-carousel": "^1.0.2",
    "react-native-draglist": "^3.6.2",
    "react-native-gesture-handler": "~2.16.1",
    "react-native-get-random-values": "^1.11.0",
    "react-native-gifted-chat": "^2.6.2",
    "react-native-google-places-autocomplete": "^2.5.6",
    "react-native-image-crop-tools": "^1.6.4",
    "react-native-keyboard-aware-scroll-view": "^0.9.5",
    "react-native-linear-gradient": "^2.8.3",
    "react-native-maps": "^1.17.3",
    "react-native-modal": "^13.0.1",
    "react-native-qrcode-svg": "^6.3.2",
    "react-native-reanimated": "~3.10.1",
    "react-native-reanimated-carousel": "^3.5.1",
    "react-native-safe-area-context": "^4.10.5",
    "react-native-screens": "3.31.1",
    "react-native-shared-element": "^0.8.9",
    "react-native-skeleton-placeholder": "^5.2.4",
    "react-native-snap-carousel": "^3.9.1",
    "react-native-svg": "^15.5.0",
    "react-native-swiper": "^1.6.0",
    "react-native-vision-camera": "^4.5.1",
    "react-native-web": "~0.19.10",
    "react-native-webview": "^13.12.2",
    "react-native-zoom-toolkit": "^3.0.0",
    "react-navigation-shared-element": "^3.1.3",
    "socket.io-client": "^4.7.5",
    "zustand": "^4.5.5"
  },
  "devDependencies": {
    "@babel/core": "^7.20.0",
    "@testing-library/jest-native": "^5.4.3",
    "@testing-library/react-native": "^12.6.1",
    "@types/jest": "^29.5.12",
    "@types/react": "~18.2.45",
    "@types/react-test-renderer": "^18.0.7",
    "jest": "^29.2.1",
    "jest-expo": "~51.0.3",
    "react-test-renderer": "18.2.0",
    "typescript": "~5.3.3"
  },
  "private": true
}
```

<SnapbackZoom>
<Image source={{ uri: 'https://(https://upload.wikimedia.org/wikipedia/commons/3/3e/Katsuyori_Shibata_November_2016.png)' }} />
</SnapbackZoom>

**Comment by Glazzes** [comment]
Hi there @Bauss9, the way you describe your issue makes me think it's because you're using relative units like `100%` , keep in mind your image is wrapped around various intermediate containers provided by this library, so that `100%` is not relative to the container you're currently working with, but the library's one, therefore you need to provide absolute values to your image for instance `{width: 375, height: 400}`.

This is my first guest, however let me know whteher it works or not for you.

**Comment by Bauss9** [comment]
Hey, thanks for the quick reply, yes what you said was indeed true, I fixed the issue now.

Do you have any solution for using this inside a Carousel?

Example:

    ```
    <Carousel
          ref={carouselRef}
          loop={false}
          width={screenWidth}
          height={IMAGE_HEIGHT}
          autoPlay={false}
          data={post.pictures}
          scrollAnimationDuration={300}
          renderItem={renderCarouselItem}
          onProgressChange={(_, absoluteProgress) => {
            progress.setValue(absoluteProgress);
            setCurrentImageIndex(Math.round(absoluteProgress));
          }}
          panGestureHandlerProps={{
            activeOffsetX: [-5, 5],
            failOffsetY: [-10, 10],
          }}
        />
```

renderCarouselItem:

```
  return (
    <GestureDetector gesture={gestureHandler}>
      <View>
        <SnapbackZoom>
        <Image 
          source={{uri: item.picture_url}}
          style={{width: 390, height: 500}}
        />
        </SnapbackZoom>
        <Animated.View style={[styles.heartContainer, heartStyle]}>
          <GradientHeart size={200} />
        </Animated.View>
      </View>
    </GestureDetector>
  );
```

Because when I zoom in, the image is contained within the carousel, not over my entire screen, which is pretty annoying. Otherwise the library is golden

**Comment by Glazzes** [comment]
@Bauss9 There are rules one can not bypass such as ScrollView and Flatlist overflow, it's simply not possible to overflow elements out of such components, components used by the carousel library you're using.

You will see this effect mostly in ScrollViews or Flatlists that take up the whole screen, for instance [this video](https://glazzes.github.io/react-native-zoom-toolkit/assets/snapbackzoom.DU-PjpnK.mp4) taken from the documentation, I used a Flatlist there, it requires more effort than it meets the eye tho.

In your situation you have two options, this assuming you list is not full screen:
- Build you own carousel using Reanimated and Gesture Handler, if it's a slider it should not take too much effort.
- Fake the effect by having an image out of the list context and mirror the state from the zoom component of the image out of the context, it can be unreliable depending how fast RN is at swapping images at the user's device.

I'd advise for the first option in most cases.

**Comment by Glazzes** [comment]
@Bauss9 Thinking more of it, there may be even be a third option, but it's difficult to tell without a visual reference of what you're building.

**Comment by Glazzes** [comment]
Hi there @ItsMeJules, If we get a bit more in depth `Gallery` is nothing but `ResumableZoom` with some custom code made for performance, being basically `ResuambleZoom` it means it can imitate `SnapbackZoom` no problem.

All you need is to set `maxScale` to one and `scaleMode` to 'bounce' likes this:
```jsx
<Gallery 
  data={data}
  renderItem={renderItem}
  maxScale={1}
  scaleMode={'bounce'} // default value
/>
```

The actual problem here is that I made `Gallery` not to overflow, I'm still thinking whether remove it entirely or create a new property for it.

If this is something you need like right now, you can patch-package this library and remove overflow style from `Gallery.tsx` and `GalleryItem.tsx` files.

**Comment by ItsMeJules** [comment]
Thanks for the response, @Glazzes.

It works, but now I have to deal with overflow issues from the carousels. (I have a shop page with multiple carousels for each product, and they are overlapping. I need to find a quick fix for this!)

**Comment by Glazzes** [comment]
@ItsMeJules It's difficult to tell what to do without a visual reference of what you're doing, I'm blindfolded.

**Comment by ItsMeJules** [comment]
@Glazzes Here some screenshots showing the issue (I couldn't upload a video sorry)

![image](https://github.com/user-attachments/assets/a4ea7605-97a2-45d2-855d-143220d89264)
![image](https://github.com/user-attachments/assets/d645e221-51a1-4a51-85e1-8e3e28b69641)
![image](https://github.com/user-attachments/assets/5269395d-667f-40bb-876d-9a4b79f2ffe1)

**Comment by Glazzes** [comment]
@ItsMeJules You've got yourself quite the use case don't you huh... I see what you want, what you attempting to do is 100% unachievable in the current state of the `Gallery`, however I think it can be done as long as `overflow` is an animatable property.

Talking of SnapbackZoom do not wrap your items on `renderItem` function with it, it's not what you think it is, it's simply makes your component zoomable, no component has the ability to bypass zIndex and overflow restrictions, I think I've misguided people with the video in the docs as I created it with those use cases in mind, however the effort to achieve that result does not rely on SnpabackZoom itself, but the way your layout is structured to handle such use case.

I can not promise a solution but it's a use case worth testing, I'll make some tests and come back to you.

**Comment by ItsMeJules** [comment]
@Glazzes thank you for your response and your time. That's what I thought.

**Comment by Glazzes** [comment]
@ItsMeJules I've made my respective research and `overflow` is a whitelist property on the UI thread so it's in fact animateable which is great, I made some testing and it's actually doable, not without effort from developer's part because galleries must "connected".

If you're eager to test you can try out this version I've compiled of the [lib](https://github.com/user-attachments/files/17085827/lib.tar.gz), to try out you need to go to this library folder in the node_modules folder and delete all the contents, then replaced by the ones in the tarball.

Here's some footage, below is the test source code I used.

https://github.com/user-attachments/assets/568fab7d-df41-47fd-8935-91a245e2e7a0

```tsx
import React, { useCallback, useEffect, useState } from 'react';
import { Image, StyleSheet, View, useWindowDimensions } from 'react-native';
import { Gallery, type SizeVector } from 'react-native-zoom-toolkit';

import {
  getAssetsAsync,
  requestPermissionsAsync,
  type Asset,
} from 'expo-media-library';
import Animated, {
  useAnimatedStyle,
  useSharedValue,
  type SharedValue,
} from 'react-native-reanimated';

type GalleryRowProps = {
  assets: Asset[];
  index: number;
  activeIndex: SharedValue<number>;
};

const GalleryRow = ({ assets, index, activeIndex }: GalleryRowProps) => {
  const { width } = useWindowDimensions();
  const gallerySize: SizeVector<number> = {
    width: (width - 32) / 2,
    height: 200,
  };

  const renderItem = useCallback(
    (item: Asset): React.ReactElement => {
      return <Image source={{ uri: item.uri }} style={gallerySize} />;
    },
    [width]
  );

  const onPinchStart = () => {
    activeIndex.value = index;
  };

  const depthStyles = useAnimatedStyle(() => {
    return {
      zIndex: index === activeIndex.value ? Number.MAX_SAFE_INTEGER : 0,
    };
  }, [activeIndex]);

  return (
    <Animated.View style={[gallerySize, depthStyles]}>
      <Gallery
        data={assets}
        renderItem={renderItem}
        maxScale={1}
        pinchCenteringMode="sync"
        onPinchStart={onPinchStart}
      />
    </Animated.View>
  );
};

const GalleryTest = () => {
  const [assets, setAssets] = useState<Asset[]>([]);
  const activeIndex = useSharedValue<number>(0);

  useEffect(() => {
    const requestAssets = async () => {
      const { granted } = await requestPermissionsAsync();
      if (granted) {
        const page = await getAssetsAsync({
          first: 5,
          mediaType: ['photo'],
          sortBy: 'creationTime',
        });

        const pageScales: SizeVector<number>[] = [];
        for (let asset of page.assets) {
          pageScales.push({ width: asset.width, height: asset.height });
        }

        setAssets(page.assets);
      }
    };

    requestAssets();
  }, []);

  return (
    <View style={styles.root}>
      <View style={styles.container}>
        {new Array(2).fill(0).map((_, index) => {
          return (
            <GalleryRow
              assets={assets}
              index={index}
              activeIndex={activeIndex}
              key={`gallery-${index}`}
            />
          );
        })}
      </View>
    </View>
  );
};

const styles = StyleSheet.create({
  root: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
  },
  container: {
    flexDirection: 'row',
    gap: 32,
    width: '100%',
    height: 200,
    backgroundColor: 'orange',
    overflow: 'visible',
  },
});

export default GalleryTest;
```

**Comment by ItsMeJules** [comment]
Thank you so much, @Glazzes! This is awesome! Although I can't use it right now (it's for production use), I'll be happy to see it in future updates.

Keep up the amazing work—this library has a solid foundation to thrive.

**Comment by Glazzes** [comment]
@ItsMeJules I've released version 3.1.0 which includes allowOverflow property, it allow the behavior seen in the video showcase above.
